### PR TITLE
Sort a few issues with NPC threat management and EP rewards

### LINKED ISSUE #133: NPC - EP Reward issue
Players are being removed from the threatmanager, but are still aggressing and killing the npc.
Thus they are not on the threatmanager on-death when EP is rewarded.

Fix/Impl detail: Collect and manage separate table for NPC to track aggressors, only remove with simple time-expiration.
(Note: timeout should be much greater than reload+cycle times)

### PR DESCRIPTION
**Pull request by LoyalServant**
1. The threat manger only removed the first entry in the list. (see ProcessHostiles)
2. The It was possible to act on a dirty list in the threatmanager.
I removed all of the logic to maintain a clean and dirty list - I feel this implementation
was overkill so this greatly simplifies things.
3. If an NPC cannot path to a player for whatever reason but that player is damaging that
NPC they will be added to a list to get half EP.

Closes: https://github.com/OpenPerpetuum/OP-Project/issues/133

### CODE REVIEW

**Comment by LoyalServant** [comment]
Don't approve this just yet - there is nothing to remove players from the 'pseudo' list.

File: src/Perpetuum/Zones/NpcSystem/Npc.cs [hunk 1]
```diff
@@ -162,6 +162,11 @@ private bool IsAttackable(Hostile hostile)
             if (hostile.unit.IsInvulnerable)
                 return false;
 
+            if (hostile.unit is Player)
+            {
+                return (hostile.unit as Player).GetLocks().OfType<UnitLock>().Where(x => x.Target.Eid == npc.Eid).FirstOrDefault() != null;
```
Comment (nref-dan) [inline src/Perpetuum/Zones/NpcSystem/Npc.cs]: You could use .Any here instead of FoD

**Review by nref-dan — COMMENTED** [review]
Everything so far looks fine, just one comment about an alternative on your target check.

**Comment by MikeJeffers** [comment]
Other contributors are welcome to take up the mantle on this task - Please see our discussion on the agreed solution on the remaining part of the implementation

**Comment by MikeJeffers** [comment]
@clouths @Ketsuekiame This is ready to be reviewed - please review!!

File: src/Perpetuum/Zones/NpcSystem/Npc.cs [hunk 2]
```diff
@@ -168,6 +168,11 @@ private bool IsAttackable(Hostile hostile)
             if (hostile.unit.IsInvulnerable)
                 return false;
 
+            if (hostile.unit is Player)
+            {
+                return (hostile.unit as Player).GetLocks().OfType<UnitLock>().Where(x => x.Target.Eid == npc.Eid).Any();
```
Comment (nref-dan) [inline src/Perpetuum/Zones/NpcSystem/Npc.cs]: Use a static cast instead of the safety cast. Better performance.

**Review by nref-dan — COMMENTED** [review]
Apart from that one comment I can't see anything particularly wrong. Best option is to get it onto something and test it out properly.

File: src/Perpetuum/Zones/NpcSystem/Npc.cs (hunk 2 above)
Comment (MikeJeffers) [reply src/Perpetuum/Zones/NpcSystem/Npc.cs]: Question: is 
```
            Player player = hostile.unit as Player;
            if (player != null)
            {
                return player.GetLocks().OfType<UnitLock>().Where(x => x.Target.Eid == npc.Eid).Any();
            }
```
better than:
```
            if (hostile.unit is Player player)
            {
                return player.GetLocks().OfType<UnitLock>().Where(x => x.Target.Eid == npc.Eid).Any();
            }
```
?

I see that `as` more efficiently evaluates type or fallsback to null (which requires that null-check), but is that the same as the alternative shown above, given that we have to null-check on that statement?

**Comment by MikeJeffers** [comment]
[NPC threatmanager bug](https://trello.com/c/G9h8SpSG/4-npc-threatmanager-bug)

**Review by MikeJeffers — APPROVED** [review]
🆗

**Pull request merged**